cxx-sensors/src/main/java/org/sonar/cxx/sensors/coverage/TestwellCtcParser.java

```diff
@@ -0,0 +1,256 @@
+  private static final int CONDS_TRUE = 2;
+  private static final int LINE_NR_GROUP = 3;
+  
+  private enum State {
+    BEGIN,
+    PARSING,
+    END;
+  }
+  
+
+  public TestwellCtcParser() {
+    // no operation but necessary for list of coverage parsers 
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void processReport(final SensorContext context, File report, final Map<String, CoverageMeasures> coverageData) {
+    LOG.debug("Parsing 'Testwell CTC++' format");
+    
+    try {
+      this.scanner = new Scanner(report).useDelimiter(SECTION_SEP);
+      this.matcher = FILE_HEADER.matcher("");
+      
+      this.state = State.BEGIN;
+      parseReportHead();
+
+      while(State.PARSING == this.state) {
+          parseUnit(coverageData);
+      }
+      
+    } catch (FileNotFoundException e) {
+      LOG.warn("TestwellCtcParser file not found '{}'", e.getMessage());
+    }
+
+  }
+  private void parseReportHead() {
+    try {
+      if (!matcher.reset(scanner.next()).find()) {
+        LOG.info("'Testwell CTC++' file section not found.");
+      } else {
+        this.state = State.PARSING;
+      }
+    } catch (NoSuchElementException e) {
+      LOG.debug("'Testwell CTC++' file section not found!");
+    }
+  }
+  
+  private void parseUnit(final Map<String, CoverageMeasures> coverageData) {
+    LOG.debug(matcher.toString());
+    
+    if (matcher.usePattern(FILE_HEADER).find(FROM_START)) {
+      parseFileUnit(coverageData);
+    } else if (matcher.usePattern(REPORT_FOOTER).find(FROM_START)) {
+      state = State.END;
+      scanner.close();
+    }
+  }
+  
+  private void parseFileUnit(final Map<String, CoverageMeasures> coverageData) {
+    LOG.debug("Parsing file section...");
+    
+    String normalFilename;
+    String filename;
+    File file;
+    
+    filename = matcher.group(1);
+    if((new File(filename)).isAbsolute() == false) {
+      normalFilename = FilenameUtils.normalize("./" + filename);
+    }
+    else {
+      normalFilename = FilenameUtils.normalize(filename);
+    }
+    file = new File(normalFilename);
+    addLines(file, coverageData);
+    matcher.reset(scanner.next());
+  }
+  
+  private void addLines(File file, final Map<String, CoverageMeasures> coverageData) {
+    LOG.debug("Parsing function sections...");
+
+    CoverageMeasures coverageMeasures = CoverageMeasures.create();
+    while (!matcher.reset(scanner.next()).usePattern(FILE_RESULT).find()) {
+      parseLineSection(coverageMeasures);
+    }
+    coverageData.put(file.getPath(), coverageMeasures);
+  }
+  
+  private void parseLineSection(CoverageMeasures coverageMeasures) {
+    LOG.debug("Found line section...");
+    
+    if (matcher.usePattern(LINE_RESULT).find(FROM_START)) {
+      addEachLine(coverageMeasures);
+    } else  {
+      LOG.warn("Neither File Result nor Line Result after FileHeader!");
+    }
+  }
+  
+  private void addEachLine(CoverageMeasures coverageMeasures) {
```

Comment:
Please reduce complexity of this method (e.g. extract methods). The Cognitive Complexity of this method is 57.

Reply:
Can you recommend a tool for Cognitive Complexity?
Which value range is good?